README.md
MeCab Chef Cookbook
=======================
Installs MeCab and its well-known dictionary.

Platforms
---------
The following platforms and versions are tested and supported using test-kitchen

* Ubuntu 14.04
* Debian 7.8
* CentOS 7.0

Attributes
-----
see [source](attributes/default.rb)

Recipes
----------
* `recipe[mecab::ipadic]` - Install MeCab and IPA Dictionary. (default)
* `recipe[mecab::jumandic]` - Install MeCab and Juman Dictionary.
* `recipe[mecab::unidic]` - Install MeCab and Unidic Dictionary.

Usage
-----
Just include `mecab` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[mecab]"
  ]
}
```

Misc
----
[GitHub Repository](http://github.com/kogecoo/chef-mecab)

License and Author
-------------------
- Author: [kogecoo](http://github.com/kogecoo)
- License: Apache License, Version 2.0 (see [LICENSE](LICENSE))


CHANGELOG.md
mecab Chef Cookbook CHANGELOG
========================

0.1.0 (Feb 19, 2015)
-----
- The initial release.


test/integration/naistjdic/naistjdic_spec.rb
require 'spec_helper'

prefix     = "/usr/local"
version    = "0.996"
dictionary = "naist-jdic"
dic_path  = "#{prefix}/lib/mecab/dic/#{dictionary}/sys.dic"

describe command('which mecab') do
  its(:stdout) { should match "#{prefix}/bin/mecab" }
end

describe file(dic_path) do
  it { should be_file }
end

describe command('mecab -v') do
  its(:stdout) { should match "#{version}" }
end

describe command('mecab -D') do
  its(:stdout) { should match (/filename:\t#{dic_path}\nversion:\t[0-9]+\ncharset:\tutf8/) }
end


test/integration/ipadic/serverspec/ipadic_spec.rb
require 'spec_helper'

# TODO: find more simple way to test.

#  Since busser-serverspec doesn't provide a feature of accessing chef attributes,
# we need to hardcode following values.
# https://github.com/test-kitchen/busser-serverspec/issues/8

prefix     = "/usr/local"
version    = "0.996"
dictionary = "ipadic"
dic_path  = "#{prefix}/lib/mecab/dic/#{dictionary}/sys.dic"


# the following are duplicated with jumandic_spec.rb and unidic_spec.rb
describe command('which mecab') do
  its(:stdout) { should match "#{prefix}/bin/mecab" }
end

describe file(dic_path) do
  it { should be_file }
end

describe command('mecab -v') do
  its(:stdout) { should match "#{version}" }
end

describe command('mecab -D') do
  its(:stdout) { should match (/filename:\t#{dic_path}\nversion:\t[0-9]+\ncharset:\tutf8/) }
end


test/integration/unidic/serverspec/unidic_spec.rb
require 'spec_helper'

prefix     = "/usr/local"
version    = "0.996"
dictionary = "unidic"
dic_path  = "#{prefix}/lib/mecab/dic/#{dictionary}/sys.dic"

describe command('which mecab') do
  its(:stdout) { should match "#{prefix}/bin/mecab" }
end

describe file(dic_path) do
  it { should be_file }
end

describe command('mecab -v') do
  its(:stdout) { should match "#{version}" }
end

describe command('mecab -D') do
  its(:stdout) { should match (/filename:\t#{dic_path}\nversion:\t[0-9]+\ncharset:\tutf8/) }
end


recipes/binary.rb
supported_versions = node['mecab']['support'].keys
version = node['mecab']['version']
mecab_dl_site = node['dl_site']['mecab']

src_filename_noext = "mecab-#{version}"
src_filename = "#{src_filename_noext}.tar.gz"
src_filepath = "#{mecab_dl_site}#{src_filename}"
copy_to = "#{Chef::Config[:file_cache_path]}/#{src_filename}"

checksum = node['mecab']['support'][version]['checksum']
checksum_type = node['mecab']['support'][version]['checksum_type']

configure_cmd = %W{
  ./configure
  #{node['mecab']['conf']['prefix'] ? "--prefix=#{node['mecab']['conf']['prefix']}" : ""}
  #{node['mecab']['conf']['charset'] ? "--with-charset=#{node['mecab']['conf']['charset']}" : ""}
  #{node['mecab']['conf']['utf8-only'] ? '--enable-utf8-only' : ''}
}.join(' ')

if not supported_versions.include?(version) then
  Chef::Application.fatal!("#{recipe_name} doesn't support the version #{version}")
end

remote_file 'download MeCab source' do
  path copy_to
  source src_filepath
  mode '0644'
  not_if 'which mecab'
  not_if { no_need_to_copy?(checksum_type, copy_to, checksum) }
  notifies :run, 'execute[make && make install MeCab]', :immediately
end

execute 'make && make install MeCab' do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  command <<-EOD
              tar -zxf #{src_filename}
              cd #{src_filename_noext}
              #{configure_cmd}
              make
              make check
              make install
            EOD
  notifies :run, 'ruby_block[edit ld.so.conf]', :immediately
end

# FIXME: Edting configuration file is not the chef way.
ruby_block 'edit ld.so.conf' do
  action :nothing
  block do
    p = "#{node['mecab']['conf']['prefix']}/lib"
    fe = Chef::Util::FileEdit.new('/etc/ld.so.conf')
    fe.insert_line_if_no_match(/p/, p)
    fe.write_file
  end
  not_if "ldconfig -p | grep #{node['mecab']['conf']['prefix']}/local/lib"
  not_if { !platform_family?('debian', 'rhel') }
  notifies :run, 'execute[ldconfig]', :immediately
end

execute 'ldconfig' do
  command 'ldconfig'
  action :nothing
end


recipes/ipadic.rb
include_recipe "mecab::binary"

supported_versions = node["ipadic"]["support"].keys
version = node["ipadic"]["version"]

src_filename_noext = "mecab-ipadic-#{version}"
src_filename = "#{src_filename_noext}.tar.gz"
src_filepath = "#{node["dl_site"]["mecab"]}#{src_filename}"
copy_to = "#{Chef::Config[:file_cache_path]}/#{src_filename}"

checksum = node['ipadic']['support'][version]['checksum']
checksum_type = node['ipadic']['support'][version]['checksum_type']

install_path_prefix = node["mecab"]["conf"]["prefix"]
mecabrc_path = "#{install_path_prefix}/etc/mecabrc"

configure_cmd = %W{
  ./configure
  --prefix=#{install_path_prefix}
  --with-charset=#{node["mecab"]["conf"]["charset"]}
}.join(" ")

if not supported_versions.include?(version) then
  Chef::Application.fatal!("This recipe for ipadic doesn't support the version #{version}")
end

remote_file copy_to do
  source src_filepath
  mode "0644"
  notifies :run, 'execute[install ipadic]', :immediately

  # Skip download/make/install if system already has IPADic regardless of its version.
  not_if { ::File.exists?("#{install_path_prefix}/lib/mecab/dic/ipadic") }
  not_if { no_need_to_copy?(checksum_type, copy_to, checksum) }
end

execute "install ipadic" do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  command <<-EOD
    tar -zxf #{src_filename}
    cd #{src_filename_noext}
    #{configure_cmd}
    make
    make install
  EOD
  notifies :run, 'ruby_block[edit dicdir path in mecabrc]', :immediately
end

# FIXME: Edting configuration file is not the chef way.
ruby_block 'edit dicdir path in mecabrc' do
  action :nothing
  block do
    fe = Chef::Util::FileEdit.new(mecabrc_path)
    fe.search_file_replace_line(/^dicdir = /, "dicdir = #{install_path_prefix}/lib/mecab/dic/ipadic")
    fe.write_file
  end
  not_if { ::File.open(mecabrc_path).read() =~ /dicdir\s+=\s+#{install_path_prefix}\/lib\/mecab\/dic\/ipadic/ }
end


recipes/unidic.rb
include_recipe "mecab::binary"

supported_versions = node["unidic"]["support"].keys
version = node["unidic"]["version"]

src_filename_noext = "unidic-mecab-#{version}_src"
src_filename = "#{src_filename_noext}.zip"
src_filepath = "http://sourceforge.jp/frs/redir.php?m=jaist&f=%2Funidic%2F58338%2F#{src_filename}"
copy_to = "#{Chef::Config[:file_cache_path]}/#{src_filename}"

checksum = node['unidic']['support'][version]['checksum']
checksum_type = node['unidic']['support'][version]['checksum_type']

install_path_prefix = node["mecab"]["conf"]["prefix"]
mecabrc_path = "#{install_path_prefix}/etc/mecabrc"

configure_cmd = %W{
  ./configure
  --prefix=#{install_path_prefix}
}.join(" ")

if not supported_versions.include?(version) then
  Chef::Application.fatal!("#{recipe_name} doesn't support the version #{version}")
end

remote_file copy_to do
  source src_filepath
  mode "0644"
  notifies :run, 'execute[install unidic]', :immediately

  # Skip download/make/install if system already has UniDic regardless of its version.
  not_if { ::File.exists?("#{install_path_prefix}/lib/mecab/dic/unidic") }
  not_if { no_need_to_copy?(checksum_type, copy_to, checksum) }
end

execute "install unidic" do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  command <<-EOD
    unzip #{src_filename}
    cd #{src_filename_noext}
    #{configure_cmd}
    make
    make install
  EOD
  notifies :run, 'ruby_block[edit dicdir path in mecabrc]', :immediately
end

# FIXME: Edting configuration file is not the chef way.
ruby_block 'edit dicdir path in mecabrc' do
  action :nothing
  block do
    fe = Chef::Util::FileEdit.new(mecabrc_path)
    fe.search_file_replace_line(/^dicdir = /, "dicdir = #{install_path_prefix}/lib/mecab/dic/unidic")
    fe.write_file
  end
  not_if { ::File.open(mecabrc_path).read() =~ /dicdir\s+=\s+#{install_path_prefix}\/lib\/mecab\/dic\/unidic/ }
end


recipes/jumandic.rb
include_recipe "mecab::binary"

supported_versions = node["jumandic"]["support"].keys
version = node["jumandic"]["version"]

src_filename_noext = "mecab-jumandic-#{version}"
src_filename = "#{src_filename_noext}.tar.gz"
src_filepath = "http://mecab.googlecode.com/files/#{src_filename}"
copy_to = "#{Chef::Config[:file_cache_path]}/#{src_filename}"

checksum = node['jumandic']['support'][version]['checksum']
checksum_type = node['jumandic']['support'][version]['checksum_type']

install_path_prefix = node["mecab"]["conf"]["prefix"]
mecabrc_path = "#{install_path_prefix}/etc/mecabrc"

configure_cmd = %W{
  ./configure
  --prefix=#{install_path_prefix}
  --with-charset=#{node["mecab"]["conf"]["charset"]}
}.join(" ")

if not supported_versions.include?(version) then
  Chef::Application.fatal!("This recipe for jumandic doesn't support the version #{version}")
end

remote_file copy_to do
  source src_filepath
  mode "0644"
  notifies :run, 'execute[install jumandic]', :immediately

  # Skip download/make/install if system already has JumanDic regardless of its version.
  not_if { ::File.exists?("#{install_path_prefix}/lib/mecab/dic/jumandic") }
  not_if { no_need_to_copy?(checksum_type, copy_to, checksum) }
end

execute "install jumandic" do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  command <<-EOD
    tar -zxf #{src_filename}
    cd #{src_filename_noext}
    #{configure_cmd}
    make install
  EOD
  notifies :run, 'ruby_block[edit dicdir path in mecabrc]', :immediately
end

# FIXME: Edting configuration file is not the chef way.
ruby_block 'edit dicdir path in mecabrc' do
  action :nothing
  block do
    fe = Chef::Util::FileEdit.new(mecabrc_path)
    fe.search_file_replace_line(/^dicdir = /, "dicdir = #{install_path_prefix}/lib/mecab/dic/jumandic")
    fe.write_file
  end
  not_if { ::File.open(mecabrc_path).read() =~ /dicdir\s+=\s+#{install_path_prefix}\/lib\/mecab\/dic\/jumandic/ }
end


recipes/naistjdic.rb
include_recipe 'mecab::binary'

supported_versions = node['naistjdic']['support'].keys
version = node['naistjdic']['version']

src_filename_noext = "mecab-naist-jdic-#{version}"
src_filename = "#{src_filename_noext}.tar.gz"
src_filepath = "#{node['dl_site']['naistjdic']}#{src_filename}"
copy_to = "#{Chef::Config[:file_cache_path]}/#{src_filename}"

checksum = node['naistjdic']['support'][version]['checksum']
checksum_type = node['naistjdic']['support'][version]['checksum_type']

install_path_prefix = node['mecab']['conf']['prefix']
mecabrc_path = "#{install_path_prefix}/etc/mecabrc"

configure_cmd = %W{
  ./configure
  --prefix=#{install_path_prefix}
  --with-charset=#{node['mecab']['conf']['charset']}
}.join(" ")

if not supported_versions.include?(version) then
  Chef::Application.fatal!("#{recipe_name} doesn't support the version #{version}")
end

remote_file copy_to do
  source src_filepath
  mode "0644"
  notifies :run, 'execute[install naist-jdic]', :immediately

  # Skip download/make/install if system already has naistjdic regardless of its version.
  not_if { ::File.exists?("#{install_path_prefix}/lib/mecab/dic/naist-jdic") }
  not_if { no_need_to_copy?(checksum_type, copy_to, checksum) }
end

execute "install naist-jdic" do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  command <<-EOD
    tar -zxf #{src_filename}
    cd #{src_filename_noext}
    #{configure_cmd}
    make
    make install
  EOD
  notifies :run, 'ruby_block[edit dicdir path in mecabrc]', :immediately
end

# FIXME: Edting configuration file is not the chef way.
ruby_block 'edit dicdir path in mecabrc' do
  action :nothing
  block do
    fe = Chef::Util::FileEdit.new(mecabrc_path)
    fe.search_file_replace_line(/^dicdir = /, "dicdir = #{install_path_prefix}/lib/mecab/dic/naist-jdic")
    fe.write_file
  end
  not_if { ::File.open(mecabrc_path).read() =~ /dicdir\s+=\s+#{install_path_prefix}\/lib\/mecab\/dic\/naist-jdic/ }
end



attributes/default.rb
default["dl_site"] = {
  "mecab"     => "http://mecab.googlecode.com/files/",
  "unidic"    => "http://sourceforge.jp/frs/redir.php?m=jaist&f=%2Funidic%2F58338%2F",
  "naistjdic" => "http://sourceforge.jp/frs/redir.php?m=iij&f=%2Fnaist-jdic%2F53500%2F"
}

default["mecab"] = {
  "version" => "0.996",
  "support" => {
    "0.996" => { "checksum" => "15baca0983a61c1a49cffd4a919463a0a39ef127", "checksum_type" => :SHA1 },
    "0.994" => { "checksum" => "9d283f9d243b1a58a2845ff60797c11a88b1f926", "checksum_type" => :SHA1 }
  },
  "conf" => {
    "prefix"   => "/usr/local",
    "charset"  => "utf8",
    "utf8-only" => true
  },
}

default["ipadic"] = {
  "version" => "2.7.0-20070801",
  "support" => {
    "2.7.0-20070801" => { "checksum" => "0d9d021853ba4bb4adfa782ea450e55bfe1a229b", "checksum_type" => :SHA1 }
  }
}

default["jumandic"] = {
  "version" => "7.0-20130310",
  "support" => {
    "7.0-20130310" => { "checksum" => "db419d23b955bfaca1e2eea421e3ffa3a1fd3f72", "checksum_type" => :SHA1 }
  }
}

default["unidic"] = {
  "version" => "2.1.2",
  "support" => {
    "2.1.2" => { "checksum" => "547ce5824429a022d6fe368af39a106c", "checksum_type" => :MD5 },
    "2.1.1" => { "checksum" => "9609881ee9c0a1d2bbc59447077863bc", "checksum_type" => :MD5 }
  }
}

default["naistjdic"] = {
  "version" => "0.6.3b-20111013",
  "support" => {
    "0.6.3b-20111013" => { "checksum" => "52238fb14d949e49a65a9dbc1f7e382c", "checksum_type" => :MD5 },
    "0.6.3-20100801"  => { "checksum" => "30af2251331b9d6141748caf033821e3", "checksum_type" => :MD5 },
    "0.6.2-20100208"  => { "checksum" => "01d806a4682da610a051480d87a2c7c2", "checksum_type" => :MD5 },
    "0.6.1-20090630"  => { "checksum" => "1a0ea25002dcf1e4b925a88e10e7d3eb", "checksum_type" => :MD5 },
    "0.6.0-20090616"  => { "checksum" => "1babf0f548b038af9d2206e6a1615a83", "checksum_type" => :MD5 },
    "0.5.0-20090512"  => { "checksum" => "90398f6be899178afea669928d5c5b70", "checksum_type" => :MD5 }
  }
}
